proof/invariant-abstract/ARM_HYP/ArchArch_AI.thy

```diff
@@ -983,24 +987,196 @@ lemma valid_global_vspace_mappings_vcpu_update_str:
   "valid_global_vspace_mappings s \<Longrightarrow> valid_global_vspace_mappings (s\<lparr>arch_state := arm_current_vcpu_update f (arch_state s)\<rparr>)"
   by (simp add: valid_global_vspace_mappings_def)
 
+lemma arm_current_vcpu_update_valid_global_vspace_mappings[simp]:
+  "valid_global_vspace_mappings (s\<lparr>arch_state := arm_current_vcpu_update f (arch_state s)\<rparr>)
+   = valid_global_vspace_mappings s"
+  by (clarsimp simp: valid_global_vspace_mappings_def global_refs_def
+              split: kernel_object.splits option.splits)
+
+crunches associate_vcpu_tcb
+  for pspace_aligned[wp]: pspace_aligned
+  and pspace_distinct[wp]: pspace_distinct
+  and zombies_final[wp]: zombies_final
+  and sym_refs_state_refs_of[wp]: "\<lambda>s. sym_refs (state_refs_of s)"
+  and valid_mdb[wp]: valid_mdb
+  and valid_ioc[wp]: valid_ioc
+  and only_idle[wp]: only_idle
+  and if_unsafe_then_cap[wp]: if_unsafe_then_cap
+  and valid_reply_caps[wp]: valid_reply_caps
+  and valid_reply_masters[wp]: valid_reply_masters
+  and valid_irq_node[wp]: valid_irq_node
+  and valid_irq_handlers[wp]: valid_irq_handlers
+  and cap_refs_in_kernel_window[wp]: cap_refs_in_kernel_window
+  and pspace_respects_device_region[wp]: pspace_respects_device_region
+  and cur_tcb[wp]: cur_tcb
+  and valid_vspace_objs[wp]: valid_vspace_objs
+  and valid_arch_caps[wp]: valid_arch_caps
+  and valid_kernel_mappings[wp]: valid_kernel_mappings
+  and equal_kernel_mappings[wp]: equal_kernel_mappings
+  and valid_asid_map[wp]: valid_asid_map
+  and valid_global_vspace_mappings[wp]: valid_global_vspace_mappings
+  and pspace_in_kernel_window[wp]: pspace_in_kernel_window
+  (wp: crunch_wps dmo_valid_irq_states device_region_dmos simp: crunch_simps)
```

Comment:
This looks like a very large crunch addition for one function. Are there not other `crunches` that could have `associate_vcpu_tcb` added to them to collapse some of these?

Reply:
I think it's nicer to concentrate them, because you get only one place to update if you change the function. One of the nice things `crunches` enabled.

Reply:
(of course if there is a second such function with a comparable set we should merge them)

Reply:
Objection your honour: `crunches` was created to do a *2D* scan of a set of properties over several functions. I guess we can place this one on the fire of aesthetic appeal... from the other perspective, if you have a bunch of functions crunching the same property, that gives you a different form of concentration. I'd ideally prefer more clustering together, especially for uninteresting things (e.g. hardly anything changes valid_asid_map/valid_global_vspace_mappings)